### PR TITLE
feat: Support search indexes on array of objects and free flow objects

### PR DESCRIPTION
**Pull request by himank**
## Describe your changes
This extends the search with following features,
- Search on array of objects (should be on top level)
- Search on free flow objects i.e. the objects with no schema (should be on top level)
- Search on primitive arrays is already supported. Primitive arrays can be on any level. This diff is adding some more tests related to it.

The following are the examples to enable this functionality,

- To enable a search index on an object without schema,

```
{
  "schema": {
    "title": "collection_a",
    "properties": {
      "field_a": {
        "searchIndex": true,
        "type": "object",
        "properties": {}
      }
    }
  }
}
```

- To enable a search index on an array of objects,
```      
{
  "schema": {
    "title": "collection_a",
    "properties": {
      "field_a": {
        "searchIndex": true,
        "type": "array",
        "items": {
          "type": "object",
          "properties": {}
        }
      }
    }
  }
}
```

- To enable a search index on an array of objects(object with schema),
```
{
  "schema": {
    "title": "collection_a",
    "properties": {
      "field_a": {
        "searchIndex": true,
        "type": "array",
        "items": {
          "type": "object",
          "properties": {
            "nested": {
              "type": "string"
            }
          }
        }
      }
    }
  }
}
```
 
 Note: `"searchIndex": true` is not needed for search indexes as default behavior is true. This is explicitly needed in case of collection schema. 
 
## Issue ticket number and link
#931

### CODE REVIEW

File: schema/fields.go [hunk 1]
```diff
@@ -337,12 +341,12 @@ type FieldBuilder struct {
...
 
-func (f *FieldBuilder) Build(forSearchIndex bool) (*Field, error) {
+func (f *FieldBuilder) Build(setSearchDefaults bool) (*Field, error) {
 	fieldType := ToFieldType(f.Type, f.Encoding, f.Format)
-	if forSearchIndex {
+	if setSearchDefaults {
 		// for search indexes, any field in schema is search indexable if it is not set explicitly.
 		// Similarly, we also tag it with sort if it is numeric.
-		if len(f.FieldName) > 0 && fieldType != ObjectType {
+		if len(f.FieldName) > 0 && (fieldType != ObjectType || (fieldType == ObjectType && len(f.Fields) == 0)) {
```
Comment (garrensmith) [inline schema/fields.go:349]: Can you move `(fieldType != ObjectType || (fieldType == ObjectType && len(f.Fields) == 0))` into its own function. Its a bit unwieldy to read

**Review by garrensmith — APPROVED** [review]
One small nit otherwise it looks good.

File: schema/fields.go (hunk 1 above)
Comment (himank) [reply schema/fields.go]: sure

**Comment by ovaistariq** [comment]
@himank could you add snippets of schema examples as well demonstrating this.

**Review by ovaistariq — DISMISSED** [review]

**Review by efirs — APPROVED** [review]

**Pull request merged**